L74-java-channel-creds.md

```diff
@@ -0,0 +1,411 @@
+```
+
+## Rationale
+
+A channel credential API is substantially harder in Java than the other
+languages as Java has multiple transports that behave radically differently.
+OkHttp uses blocking I/O with a byte[]-backed Buffer whereas Netty uses
+nonblocking I/O with a reference-counted, ByteBuffer-backed ByteBuf with an
+event loop. Go and C use a code-based API where credentials implement an
+interface and use a single I/O model. It is not practical to have one
+code-based API to support both OkHttp and Netty, as it would have a
+considerable amount of complexity in the implementations, be hard to optimize,
+and have a wide public API surface. Instead, Java needs a data-based API, where
+configuration values are communicated and each transport can have its own
+implementation. A data-based API can also be used by InProcess and Binder
+transports.
+
+There needs to be a `TlsChannelCredential` which is able to work with both
+OkHttp and Netty transports. This credential needs to live in io.grpc and so
+cannot depend on OkHttp nor Netty. Since it is not feasible for the credential
+to provide the TLS implementation, it will provide configuration. Since new
+configuration will be added over time, there needs to be a way for a transport
+to know that it understands the important parts of the configuration.
+
+To support transport credentials like ALTS and TLS-offloading, there still
+needs to be a way to provide implementations. However, these implementations
+_can_ depend on the specific transport being used. Just the users of those
+implementations should avoid depending _explicitly_ on a particular transport.
+
+The API surface should handle generic and transport-specific credentials in the
+same manner, so the user’s code should not care about the implementation
+details.
+
+It would be possible to use a versioning scheme for `TlsChannelCredentials`.
+Version 1 could be all defaults. Version 2 could have file-based client
+certificates and ca certificates for trust. Version 3 could include ciphers,
+KeyManagers, TrustManagers, and session cache configuration. While versions
+safely allow adding features in the future and is simpler than the Features
+enum approach, it means a consumer that only supports a subset of features at a
+version must check _all_ the features at a particular version to see if they
+are set. Keeping track of which features were available at each version can be
+error-prone. It's also hard to produce a user-meaningful error message, so they
+are aware of which specific configuration is causing the incompatibility.
+
+## Implementation
+
+ejona86 will implement. TLS will initially be bare-bones, with just the
+compatibility checker. Over the long-term TLS will be fleshed out with
+additional configuration. ServerCredentials will be implemented shortly after
+ServerCredentials.
+
+## Open issues (if applicable)
+
+Do we put the constructor in ManagedChannel instead of ManagedChannelBuilder?
```

Comment:
API review notes about this API:
 * io.grpc.Grpc was the preferred place for the builder-creation method, to avoid the current ManagedChannelBuilder issues and avoid the potential for the same issue cropping up with ManagedChannel
 * We didn't get to a point of really choosing a name for the method though. Grpc.newManagedChannelBuilderForTarget is pretty obnoxious. There was a desire expressed to keep "for target" in the name, as we will have "for address" as well. "Managed" may be easy to remove.